libutils/glob_lib.c

```diff
@@ -0,0 +1,512 @@
+
+static void ExpandBraces(const char *const pattern, StringSet *const expanded)
+{
+    assert(pattern != NULL);
+    assert(expanded != NULL);
+
+    char *const start = SafeStringDuplicate(pattern);
+
+    /* First we will split the pattern into three parts based on the curly
+     * brace delimiters. We look for the shortest possible match. E.g., the
+     * string "foo{{bar,baz}qux,}" becomes "foo{", "bar,baz", "qux,}".
+     */
+    char *left = NULL, *right = NULL;
+    for (char *ch = start; *ch != '\0'; ch++)
+    {
+        if (*ch == '{')
+        {
+            left = ch;
+        }
+        else if (left != NULL && *ch == '}')
+        {
+            right = ch;
+            break;
+        }
+    }
+
+    /* Next we check if base case is reached. I.e., if there is no curly brace
+     * pair to expand, then add the pattern itself to the result.
+     */
+    if (right == NULL)
+    {
+        StringSetAdd(expanded, start);
+        return;
+    }
+    *left = '\0';
+    *right = '\0';
+
+    /* Next we split the middle part (between the braces) on the comma
+     * delimiter. E.g., the string "bar,baz" becomes "bar", "baz".
+     */
+    char *middle = left + 1, *end = right + 1;
+    Seq *const split = StringSplit(middle, ",");
+
+    /* Lastly we combine the three parts for each split element, and
+     * recursively expand subsequent curly braces.
+     */
+    const size_t len = SeqLength(split);
+    for (size_t i = 0; i < len; i++)
+    {
+        middle = SeqAt(split, i);
+        char *const next = StringConcatenate(3, start, middle, end);
+        ExpandBraces(next, expanded);
+        free(next);
+    }
+
+    free(start);
+    SeqDestroy(split);
+}
+
+#ifdef WITH_PCRE2
+
+/**
+ * Translate the square bracket part of a shell expression (glob) into a regular
+ * expression.
+ */
+static size_t TranslateBracket(
+    const char *const pattern, const size_t n, size_t left, Buffer *const buf)
+{
+    size_t right = left;
+    if (right < n && pattern[right] == '!')
+    {
+        right += 1;
+    }
+    if (right < n && pattern[right] == ']')
+    {
+        right += 1;
+    }
+    while (right < n && pattern[right] != ']')
+    {
+        right += 1;
+    }
+    if (right >= n)
+    {
+        /* There was an opening bracket, but no closing one. Treat it as part
+         * of the filename by escaping it.
+         */
+        BufferAppendString(buf, "\\[");
+    }
+    else
+    {
+        char *stuff = SafeStringNDuplicate(pattern + left, right - left);
+        if (StringContains(stuff, "--"))
+        {
+            // Escape backslashes and hyphens for set difference (--).
+            // Hyphens that create ranges shouldn't be escaped.
+            Seq *const chunks = SeqNew(1, free);
+            size_t middle = (pattern[left] == '!') ? left + 2 : left + 1;
+            while (true)
+            {
+                middle = strcspn(pattern + middle, "-");
+                if (middle >= strlen(pattern + middle))
```

Comment:
@larsewi On these 2 lines of code, `pattern + middle` is different.

The problem is that `middle` is an index in reference to what was sent in as the argument.
In this case `pattern + middle` is the string we are referring to, the return value of `strcspn` should be related to this.
But since you assign to `middle`, we don't know what `pattern + middle` was any more.

```suggestion
                const char *const middle_string = pattern + middle.
                middle = strcspn(middle_string, "-");
                if (middle >= strlen(middle_string))
```

When using these C functions which work on indices and strings (pointers), it's really important to keep track of which string (pointer) an index is referring to.

(I haven't reviewed the rest of the code below).

Reply:
Thanks!